crates/biome_js_analyze/src/analyzers/style/use_consistent_array_type.rs

```diff
@@ -0,0 +1,485 @@
+    /// ### Invalid
+    /// ```ts,expect_diagnostic
+    /// let invalid: Array<foo>;
+    /// ```
+    ///
+    /// ```ts,expect_diagnostic
+    /// let invalid: Promise<Array<string>>;
+    /// ```
+    ///
+    /// ```ts,expect_diagnostic
+    /// ```
+    ///
+    /// ### Valid
+    /// ```ts
+    /// const valid: Array<string | number> = ['a', 'b'];
+    /// const valid: Array<{ prop: string }> = [{ prop: 'a' }];
+    /// const valid: Array<() => void> = [() => {}];
+    /// const valid: MyType[] = ['a', 'b'];
+    /// const valid: string[] = ['a', 'b'];
+    /// const valid: readonly string[] = ['a', 'b'];
+    /// ```
+    ///
+    /// ## Options
+    /// The rule provides two options that help to specify what type of array declarations to use.
+    ///
+    /// Default: "simple"
+    ///
+    /// ```json
+    /// {
+    ///     "//": "...",
+    ///     "options": {
+    ///         "consistentArrayType": "simple" | "generic"
+    ///     },
+    /// }
+    /// ```
+    ///
+    /// By default, all array declarations will be converted to `T[]` or `readonly T[]`, which it means `simple`,
+    /// or if the options is set to the "generic", that all will converted to `Array<T>` or `ReadonlyArray<T>`.
+    ///
+    pub(crate) UseConsistentArrayType {
+        version: "1.4.2",
+        name: "useConsistentArrayType",
+        recommended: false,
+        fix_kind: FixKind::Unsafe,
+    }
+}
+
+declare_node_union! {
+    pub(crate) Query = TsTypeAnnotation | TsReferenceType
+}
+
+#[derive(Debug, Copy, Clone)]
+enum TsArrayKind {
+    /// `T[]`
+    Simple,
+    /// `Array<T>`
+    GenericArray,
+    /// `readonly T[]`
+    Readonly,
+    /// `ReadonlyArray<T>`
+    ReadonlyGenericArray,
+}
+
+impl Rule for UseConsistentArrayType {
+    type Query = Ast<Query>;
+    type State = AnyTsType;
+    type Signals = Option<Self::State>;
+    type Options = ConsistentArrayType;
+
+    fn run(ctx: &RuleContext<Self>) -> Option<Self::State> {
+        let query = ctx.query();
+        let options = ctx.options();
+
+        match query {
+            Query::TsTypeAnnotation(ty) => {
+                if options.consistent_array_type == "simple" {
+                    return None;
+                }
+
+                let ty: AnyTsType = ty.ty().ok()?;
+                match get_array_kind_by_any_type(&ty) {
+                    Some(array_kind) => transform_array_type(ty, array_kind),
+                    _ => None,
+                }
+            }
+            Query::TsReferenceType(ty) => {
+                if options.consistent_array_type == "generic" {
+                    return None;
+                }
+
+                match get_array_kind_by_reference(ty) {
+                    Some(array_kind) => {
+                        let type_arguments = &ty.type_arguments()?;
+                        convert_to_array_type(type_arguments, array_kind)
+                    }
+                    _ => None,
+                }
+            }
+        }
```

Comment:
If we directly query `AnyTsType`, we have to add:
```suggestion
            _ => None,
        }
```

Reply:
Don~